### PR TITLE
feature: AdExRecoveryDao

### PR DESCRIPTION
**Pull request by samparsky**
Implements AdExRecoveryDao contract

### CODE REVIEW

File: contracts/AdExRecoveryDAO.sol [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+pragma solidity ^0.6.12;
+pragma experimental ABIEncoderV2;
+
+import "./Identity.sol";
+
+library RecoveryRequestLibrary {
+    struct RecoveryRequest {
+        address identity;
+        address newUserAddress;
```
Comment (Ivshti) [inline contracts/AdExRecoveryDAO.sol:11]: should be `newWalletAddress` -

File: contracts/AdExRecoveryDAO.sol [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        address proposer;
```
Comment (Ivshti) [inline contracts/AdExRecoveryDAO.sol:12]: proposer is not needed

**Review by Ivshti — CHANGES_REQUESTED** [review]
any `proposer` should be able to propose or finalize a recovery, as such `request.proposer` is not needed
It seems to me that `request.timestamp` is not needed as well

File: contracts/AdExRecoveryDAO.sol [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        proposers[proposer] = true;
+        emit LogAddProposer(proposer, now);
+    }
+
+    /**
+     * @notice Only the admin is allowed to remove a proposer
+     * @param proposer The address of the proposer
+     */
+    function removeProposer(address proposer) external {
+        require(msg.sender == adminAddr, 'ONLY_ADMIN_CAN_REMOVE_PROPOSER');
```
Comment (rori4) [inline contracts/AdExRecoveryDAO.sol:72]: why not create a modifier `onlyAdmin` and use that?

File: contracts/AdExRecoveryDAO.sol [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        delete proposers[proposer];
+        emit LogRemoveProposer(proposer, now);
+    }
+
+    /**
+     * @notice Only proposers are allowed to a propose recovery request
+     * @param request The details of the recovery request
+     */
+    function proposeRecovery(RecoveryRequestLibrary.RecoveryRequest memory request) external {
+        require(proposers[msg.sender] == true, 'ONLY_WHITELISTED_PROPOSER');
```
Comment (rori4) [inline contracts/AdExRecoveryDAO.sol:82]: same for this require the one that is used below in `finalizeRecovery`. Btw this sends "ONLY_WHITELISTED_PROPOSER" the other one is "ONLY_WHITELISTED_PROPOSERS".

File: contracts/AdExRecoveryDAO.sol [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        delete recovery[recoveryId];
+        emit LogCancelRecovery(recoveryId, msg.sender, now);
+    }
+    
+    /**
+     * @notice Only Admin can replace themself
+     * @param newAdmin The address of the new admin
+     */
+    function changeAdmin(address newAdmin) external {
+        require(msg.sender == adminAddr, 'ONLY_ADMIN_CAN_CALL');
```
Comment (rori4) [inline contracts/AdExRecoveryDAO.sol:140]: same here. you repeat that in `changeRecoveryDelay`

File: contracts/AdExRecoveryDAO.sol [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        require(newAdmin != adminAddr, 'INVALID_NEW_ADMIN');
```
Comment (IvoPaunov) [inline contracts/AdExRecoveryDAO.sol:141]: maybe `require(admin != address(0), 'INVALID_ADMIN')` or it is on purpose without that check.

File: contracts/AdExRecoveryDAO.sol (hunk 3 above)
Comment (samparsky) [reply contracts/AdExRecoveryDAO.sol:72]: Under the Code style https://github.com/AdExNetwork/adex-protocol-eth#code-style-and-design-principles  we don't use modifiers

Comment (rori4) [reply contracts/AdExRecoveryDAO.sol:72]: Ou ok. Then nevermind the rest of the comments as well

**Comment by Ivshti** [comment]
@samparsky when we merge v5, conflicts here should be resolved so we can merge this too

**Review by Ivshti — APPROVED** [review]

**Comment by Ivshti** [comment]
closing in favor of https://github.com/AdExNetwork/adex-protocol-eth/blob/identity-v5.2/contracts/wallet/QuickAccManager.sol

**Pull request closed without merging**